### PR TITLE
Major API changes

### PR DESCRIPTION
**Pull request by ner24**
The old API worked fine. However, I am not sure if it guaranteed safe casts between different CAN message types. Only the byte alignment for the data in the library can_frame struct was specified. The padding around can_id and can_dlc were not specified. So, even though this didn't cause an issue, I still want to change it to try and guarantee a safe cast. So, I have switched to using a union of all the possible message types. This union is wrapped in another union with the can_frame buffer. There is padding in between the align the unions correctly and is worked out using a offsetof macro defined by the c standard. This draft does not increase the number of memory copies compared to the old method (which had no memory copies outside of ones in the library).

### CODE REVIEW

**Comment by ner24** [comment]
The draft is in translator2024. I have not tested this yet.

**Comment by ner24** [comment]
Also this includes a clean up of the CanApi header

**Comment by ner24** [comment]
Test results to are in translator2024/generatedAPIs. Syntax is fine but behaviour not tested yet

**Comment by ner24** [comment]
The changes have now been verified with telemetry and a test board which sends driver controls messages using the same CANTranslator api.

**Comment by ner24** [comment]
Side note this has also verified the molex to m12 adapters (at least the CAN-h and CAN-l connections).

File: config/CANBusData(saved201022)Modified.xlsm
Comment (owentmfoo) [inline config/CANBusData(saved201022)Modified.xlsm:1]: not needed anymore? Replaced by CANBusDataWithEndianness?

File: config/CANBusConfig.xlsm
Comment (owentmfoo) [inline config/CANBusConfig.xlsm:1]: not needed anymore? Replaced by CANBusDataWithEndianness?

File: translator/oldAPIs/CanApiv02.cpp
Comment (owentmfoo) [inline translator/oldAPIs/CanApiv02.cpp:1]: remove unused code?

File: translator2024/CanApiGen.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    processMsgDeclarationsStream.write("\t\tvoid processAll(CANHelperBuffer& msg);\n")
+    processMsgDeclarationsStream.write("#endif\n")
+
+    #add union
+    headerStream.write("\tunion CastedCANPayload {\n" + unionDeclarationStream.getvalue() + "\t};\n")
+    unionDeclarationStream.close()
+    headerStream.write("}\n") #close Messages namespace
+
+    #add helper buffer union and MsgHandler class
+    headerStream.write(""" namespace CANHelper {
```
Comment (owentmfoo) [inline translator2024/CanApiGen.py]: extra space?

File: translator2024/CanApiDraft.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+#ifndef CanApiv04_INCLUDE_GUARD
+#define CanApiv04_INCLUDE_GUARD
+#include "mcp2515.h" //for certain datatypes used in this file
+
+namespace CANHelper::Messages //union containing all messages is at bottom
+{
+#ifdef USE_MSG_Telemetry_TimeAndFix
+#define CAN_ID_Telemetry_TimeAndFix 0x0F6
+#define CAN_DLC_Telemetry_TimeAndFix 8
+	struct Telemetry::_TimeAndFix { //Python code will remove underscores from item names (just havnt changed them here)
```
Comment (owentmfoo) [inline translator2024/CanApiDraft.hpp]: why the comment instead of removing the underscore here?

**Review by owentmfoo — APPROVED** [review]

File: config/CANBusConfig.xlsm
Comment (ner24) [reply config/CANBusConfig.xlsm:1]: resolved with commit 0781f45012ffe56b7050ca6fc5ba70015ea38047

File: config/CANBusData(saved201022)Modified.xlsm
Comment (ner24) [reply config/CANBusData(saved201022)Modified.xlsm:1]: resolved with commit https://github.com/DUEM/CANTranslator/commit/0781f45012ffe56b7050ca6fc5ba70015ea38047

File: translator/oldAPIs/CanApiv02.cpp
Comment (ner24) [reply translator/oldAPIs/CanApiv02.cpp:1]: resolved with https://github.com/DUEM/CANTranslator/pull/10/commits/d126975ef66f13eb438293ad09f4ccb1a2b70293

File: translator2024/CanApiGen.py (hunk 1 above)
Comment (ner24) [reply translator2024/CanApiGen.py]: resolved with https://github.com/DUEM/CANTranslator/pull/10/commits/4b1304bb9474a0d4a9ac2611ee3e728068bb0584

File: translator2024/CanApiDraft.hpp (hunk 2 above)
Comment (ner24) [reply translator2024/CanApiDraft.hpp]: resolved with https://github.com/DUEM/CANTranslator/pull/10/commits/7991e6e98b24f421fc604c4e7814634c0dc08116

**Pull request merged**